### ISSUE TITLE
Optimize phasing pipeline for scalability.

### ISSUE THREAD

**Issue by samuelklee** [broadinstitute/lrma-aou1-panel-creation#38]
Now that we've reconfigured and refactored the phasing pipeline to improve SV performance on 10Mbp, we probably need to make additional changes to improve scalability for whole chromosome/genome runs.

Here are a few submissions to reference:

- old pipeline, chr6, ~$170, HiPhase 16G, Shapeit4 96G (SV) / 128G (scaffold), scaffold took **~120 hours**: https://app.terra.bio/#workspaces/allofus-drc-wgs-lr-prod/AoU_DRC_WGS_LongReads_PacBio/job_history/ebfa6196-d25c-4830-acc3-23bd77dd691b
- new pipeline, chr6, **~$400** + ~$30 + $???, HiPhase 16G (1st run, few samples failed) -> 32G, Shapeit4 32G (2nd run, failed) -> 128G (3rd run, still running after 13 hours): https://app.terra.bio/#workspaces/allofus-drc-wgs-lr-prod/AoU_DRC_WGS_LongReads_PacBio%20PAPER%20COPY/job_history/9cd3ef6f-1f4c-4ba7-8f67-ecf76f91c391, https://app.terra.bio/#workspaces/allofus-drc-wgs-lr-prod/AoU_DRC_WGS_LongReads_PacBio%20PAPER%20COPY/job_history/72113057-45f0-4e02-b978-316ab261d2d0, and https://app.terra.bio/#workspaces/allofus-drc-wgs-lr-prod/AoU_DRC_WGS_LongReads_PacBio%20PAPER%20COPY/job_history/dea0b0fd-fa39-400e-8a10-655be8c99eb2 (still running)

I think I will leave this last run going, but probably we should make some of the following improvements and rerun:

- Reexamine necessity of BAM/VCF subsetting. BAM subsetting by region was added to the new workflow to facilitate running on 10Mbp. I believe VCF subsetting by sample was also added, but I'm not sure this was necessary, as it looks like HiPhase can take as input a multisample VCF and will only retrieve the samples needed. I still think we need to be able to run the workflow on arbitrary genomic regions, but let's take out subsetting and reduce the number of per-sample jobs where possible. I hope this will also improve call caching---I was getting a lot of misses and/or slow caches, and I suspect having a lot of per-sample jobs exacerbated this.
- Examine possibility of batching HiPhase. It looks like HiPhase can run on multiple samples in a single invocation. If it doesn't make too much of a difference, this is probably lower priority than removing extraneous per-sample subsetting jobs.
- Do not log HiPhase verbosely. On the chr6 runs, it looks like this generates a 100Mb log per sample.
- Shard Shapeit4. This is a **must do**, IMO. Several days is too long for any job and the tool is explicitly designed to be run on genomic chunks that are then subsequently ligated.

I think some of the slowness and inefficiency is due to Cromwell/Terra (e.g., slow or missed caching), but we can certainly clean things up on our end.

@hangsuUNC @SHuang-Broad @kvg take note. In particular, I'm curious if @SHuang-Broad has more insight about making things more cache friendly.

**Comment by hangsuUNC** [comment]
Cool, Thanks for the experiments! I guess this will be my todo list:
1) remove BAM and VCF subsetting.
2) Run Hiphase in whole genome (?), remove verbose
3) Shard Shapeit4 and ligate

**Comment by samuelklee** [comment]
Just to clarify, we should remove subsetting only as it makes sense to! Now that I'm looking at the VCF sizes (~60GB for the joint short callset), it seems that maybe keeping VCF subsetting by sample during HiPhase is sensible. In any case, ideally this is streamed, but it looks like that's not the case now in the SubsetVCF task.

I probably wouldn't sweat subsetting too much during Shapeit4 at first, but if it's easy enough to add it now, go ahead!

**Comment by hangsuUNC** [comment]
I tried to add -`-sample-name ~{samplename}`  to hiphase and give the joint small and SV vcf to hiphase directly. The output of small and sv is a multiple-sample vcf, which means we still need to split these vcf into single-sample vcf and merge them. Here is the test run:https://app.terra.bio/#workspaces/broad-firecloud-dsde/lrma-aou1-panel-creation-hprc-only/job_history/21097155-6346-44b7-82cf-26f41f498d5e

**Comment by SHuang-Broad** [comment]
When doing the sharding of the BAMs, I noticed too that Cromwell didn't seem to call-cache correctly. Unfortunately, I didn't dig into why. Generally, I find a combination of 

- sharding widely with 
- a lot of entitities to be sharded widely, 

creates issues, like you observed.

Based on my understanding of the current (whole genome) phasing pipeline, sharding of the BAM happens right before the HiPhase step, and each BAM is sharded per chromosome. A quick sketch gives us ~1000(samples)*23(chromosomes), which is quite high. So the issue related to the physical phasing now, is that it probably is too-finely sharded.

Since the goal is to make the physical phasing step efficient, I can imagine a few different experiments to run before making a decision (some data may already be available):

- how long does HiPhase run on a lrWGS BAM? If only a few hours per sample, it probably makes sense to split the "joint" VCFs (small and SV) by sample (once, though not cheap). We don't touch the BAMs in this approach. And HiPhase works on a single sample in each instance. We can even batch the samples into groups (say 10), if the total number of samples is high enough to worry us about the width of sharding. The risk here, though, is if HiPhase runs longer than a handful of hours, preemption could kick in. We can play some tricks in trying to further help HiPhase to parallelize inside a task, but we probably need to benchmark that first. 
- how long does HiPhase run on a cohort of lrWGS samples, but restricted to only a chromosome? We'll need to subset the two VCFs by chromosome (cheap), and shard the BAM (expensive and very wide scatter). The plus for going with this approach is the limited number of HiPhase runs (# of chromosomes). The minuses are: 1) HiPhase potentially being slow when given 1000 samples 2) Cromwell issues from the wide scatter (need for call-caching also means storage costs).

My guess is now 1) might be better, if HiPhase can finish a lrWGS sample in a reasonable amount of time.

-----------

On sharding ShapeIt4 then ligating back, I agree that we definitely should do it.
The UKBB [phasing report](https://biobank.ndph.ox.ac.uk/showcase/ukb/docs/PhasingUKB200k_report_SHAPEIT.pdf) gives guidance on how ShapeIt5 was used that way.

**Comment by hangsuUNC** [comment]
The script with sharded Shapeit 4 is here: https://github.com/broadinstitute/lrma-aou1-panel-creation/blob/hs_filter/wdl/methods/phasing/PhysicalAndStatisticalPhasing.wdl
The job is here: https://app.terra.bio/#workspaces/broad-firecloud-dsde/lrma-aou1-panel-creation-hprc-only/job_history/016072d7-4c06-4d76-a713-3c1e67acfb19

**Comment by hangsuUNC** [comment]
ligated shapeit4 change the final switch error a little but do not have a significant impact.  Here is the three samples comparison:
<img width="466" alt="image" src="https://github.com/user-attachments/assets/57e41f86-a155-4c37-b85b-b64119647e20">

The control run: https://app.terra.bio/#workspaces/broad-firecloud-dsde/lrma-aou1-panel-creation-hprc-only/job_history/2b59a709-e396-40c1-b225-6bbdb665f96f

The ligated run: https://app.terra.bio/#workspaces/broad-firecloud-dsde/lrma-aou1-panel-creation-hprc-only/job_history/1d2f1230-ab2c-4c40-9251-d7009f9f01c3

**Comment by samuelklee** [comment]
Also, I see a drop in SV precision from ~82% to ~71%. Not sure if this is due to Shapeit4 reproducibility (and also the shard size here is quite small compared to the amount of overlap), but please keep an eye on it.

**Comment by samuelklee** [comment]
Actually, looks like num_threads = 1 for Shapeit4 here, so it might be due to the overlap vs. shard size ratio. Might be worth experimenting further.

**Comment by hangsuUNC** [comment]
Yes, adding switch error rate to the SummarizeEvaluation task would be great!

**Comment by samuelklee** [comment]
@hangsuUNC has already made some nice progress towards using `bcftools +split` and optimizing subsetting in her `hs_filter` branch. For a cost baseline to compare against, I ran `main` on HPRC-only 10Mbp (https://app.terra.bio/#workspaces/broad-firecloud-dsde/lrma-aou1-panel-creation-hprc-only/job_history/7d260869-769c-43ee-8eb3-d1019aae7753), which gave the following costs:
![cost](https://github.com/user-attachments/assets/c98c2c6b-25a6-47fb-b387-e41aa1167c37)

We already tracked down the the split-by-sample tasks used a very large VM (64G), which probably accounted for most of the cost of that inadvertently expensive chr6 run. But hopefully using `bcftools +split` gives further improvement beyond simply reducing this VM size.

**Comment by hangsuUNC** [comment]
Testing on HPRC+AoU 1074 samples on chr1:100-110 10Mb region has finished. The baseline workflow (main branch) cost is 
$93.89 (https://app.terra.bio/#workspaces/allofus-drc-wgs-lr-prod/AoU_DRC_WGS_LongReads_PacBio%20PAPER%20COPY/job_history/78119dfb-966d-4de9-b8ed-0572d519fdbe). The hs_filter branch cost is $13.85 (https://app.terra.bio/#workspaces/allofus-drc-wgs-lr-prod/AoU_DRC_WGS_LongReads_PacBio%20PAPER%20COPY/job_history/41df5e00-4b15-485b-abf0-dd7cc71f3eca).

**Comment by samuelklee** [comment]
This is great! I'm guessing that main run still used the 64GB VMs? (If not, even better!) If so, it might be good to report the TotCost and StaticCost (where the latter is supposed to reflect what the workflow would've cost if all VMs were sized such that the most expensive shards of each task would've been accommodated).

**Comment by hangsuUNC** [comment]
The precision of Small and SVs gets even better after bcftools split and shapeit4 sharded:
<img width="711" alt="image" src="https://github.com/user-attachments/assets/2ee2d536-aace-4628-adde-41fd3c347242">

**Comment by hangsuUNC** [comment]
slightly increase switch/flip error:
<img width="523" alt="image" src="https://github.com/user-attachments/assets/b7fc947f-cd72-48b1-89a9-36c6f7c4d565">

**Comment by hangsuUNC** [comment]
Cost comparison:
Baseline:
![baseline_monitoring cost](https://github.com/user-attachments/assets/a6c3b135-dc34-4967-94ca-21463dc23530)
hs_filtering:
![hs_filtering_monitoring cost](https://github.com/user-attachments/assets/6950fad9-55b2-448e-a359-57e87b9a1254)

**Comment by samuelklee** [comment]
For after the weekend:

The HPRC+AoU1 chr6 run (https://app.terra.bio/#workspaces/allofus-drc-wgs-lr-prod/AoU_DRC_WGS_LongReads_PacBio%20PAPER%20COPY/job_history/fc3f241d-6d28-4426-8593-578412b7f763) is still failing, and perhaps could use further optimization.

In particular, please examine the split tasks and ensure that VMs are appropriately sized and utilized in terms of disk, memory, and CPU. I would also take a look at the breakdown between the split command and the serial/looped tabix indexing and optimize accordingly.

Similarly, it looks like HiPhase is occasionally hitting OOM. (Not sure why automatic retrying isn’t kicking in, but I wouldn’t rely on it.) We can either size VMs more conservatively or explore the impact of filtering before HiPhase, as filed in #40.

**Comment by samuelklee** [comment]
Still making progress, but the latest submission (https://app.terra.bio/#workspaces/allofus-drc-wgs-lr-prod/AoU_DRC_WGS_LongReads_PacBio%20PAPER%20COPY/job_history/f4abedea-c158-4065-b45d-780bc908c008, in which HiPhase memory was bumped from 16GB to 32GB) failed after ~33 hours due to a Shapeit4 shard running OOM with 64GB.

A few comments:

- Upping HiPhase memory to 32GB is a bit heavy handed and we will pay a ~1k multiplier. Examining some of the monitoring logs shows that some samples use as few as 5GB. Perhaps use a simple scaling rule depending on the size of the VCFs? Ideally, it would be easy to figure out the appropriate sizing when we move to whole-genome runs.
- A large amount of time is still being devoted to relatively straightforward VCF I/O; e.g., ~3/6/11 hours for subsetting/splitting/merging short variants and ~3 hours for filter/concat. This suggests: 1) we should try to use BCF where possible, 2) we should ensure threading is utilized where possible, 3) we may want to do a hierarchical merge, and 4) we might indeed want to filter beforehand as suggested in #40.
- We should probably use variant count to determine the Shapeit4 shard regions, as is done e.g. in the WARP joint-calling pipeline (using a Picard tool, IIRC).


As always, we should use the monitoring logs to guide us and do test runs of tasks in single VMs. We should have a rough idea of how long a run will take and how much the most expensive tasks will cost before we kick it off.

I think we can aim for a whole-genome run that completes in under 24 hours, with most of the expensive compute and processing time being devoted to HiPhase and Shapeit4.

**Comment by SHuang-Broad** [comment]
> Similarly, it looks like HiPhase is occasionally hitting OOM. (Not sure why automatic retrying isn’t kicking in, but I wouldn’t rely on it.) We can either size VMs more conservatively or explore the impact of filtering before HiPhase, as filed in https://github.com/broadinstitute/lrma-aou1-panel-creation/issues/40.

It may be worth it now to perform monitoring with Beri's monitoring system, which is much more powerful than the script approach (it can be merged with workflow metadata to even permit some modeling).

**Comment by samuelklee** [comment]
@hangsuUNC has made a lot of progress, but after looking at runtime numbers some more my feeling is that things are still a bit unwieldy (and will be even more so for Phase 2). I am also finding the same downstream for GLIMPSE (where I am still running per-chromosome, with very small batches of 5-10 samples---probably we should eventually use the same sharding scheme and run over the whole cohort as with Shapeit4).

I think the obvious solution, as long as we and our downstream users are OK with it, is to aggressively drop short variants. Our current panel, after dropping singletons, contains ~2.3M SNVs on chr6 alone. Compare with the ~57k SNVs used *genome-wide* in the 888 sample 1kG panel in https://www.medrxiv.org/content/10.1101/2023.12.20.23300308v1 (randomly subsampled from the ~702k SNVs intersecting with those present in UKBB to "facilitate benchmarking").

It seems that the general sentiment from the working group when asked about the short variants has always been "they would be nice to have," but I don't know if we've ever had a serious cost-benefit discussion. Given that the eQTL folks seem perfectly content to initially move forward with Paragraph-based genotypes (which only yield SVs), it seems a little silly to hamstring everything by a factor of nearly 1000x if no one is going to be interested in the majority of those genotypes.

I would suggest that we do a quick preliminary analyses to see the short-variant count and the corresponding downstream runtime and accuracy in padded windows around SVs, which have presumably already been generated in the course of the recent hapestry work to integrate short variants. My guess is that SV performance will not drop too much, especially given prior results in the KAGE papers and the fact that conventional imputation shard sizes are as small as a few Mb.

@kvg @hangsuUNC @fabio-cunial thoughts? If we're OK to move forward, I might as Fabio for a BED file with those padded windows for the TPR = 0.7 v1 callset, if such a thing already exists. Then I can kick off some quick chr6 HPRC+AoU1 evaluations today.

**Comment by hangsuUNC** [comment]
Agree. Reading HLA imputation related work recently (most small variants) and they do restricted Per-variant QCs before going to the phasing and imputation step. They first deduplicate variants on the basis of genomic position and alleles by removing the ones with higher missingness. Then the following filters include: 1) variants with high missingness (>0.01), 2) variants demonstrating a significant deviation from HWE; 3) variants with very low MAF.
However, we should also consider in MHC region, they have highly variable allele frequency (AF) across populations and expected HWE deviation in the MHC variants due to nature selection.

**Comment by samuelklee** [comment]
On 2x10Mb Shapeit4 shards of chr6, AF>=0.01 and SV 10kb windows yielded <1hr runtimes with reasonable VMs. Recall of short variants decreased from 90+% to ~60%, but all other metrics stayed about the same.

with filtering and windowing: https://app.terra.bio/#workspaces/allofus-drc-wgs-lr-prod/AoU_DRC_WGS_LongReads_PacBio%20PAPER%20COPY/job_history/21a8aef6-4721-4b95-8d75-91f1038b2115
baseline: https://app.terra.bio/#workspaces/allofus-drc-wgs-lr-prod/AoU_DRC_WGS_LongReads_PacBio%20PAPER%20COPY/job_history/6e2c5ed2-a102-466e-9dbc-8eae7bae1021
